File: LeetCode75/Linked_List_Cycle_II/Linked_List_Cycle_II.py
# 142. Linked List Cycle II Python

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tmp = head
        index = 0
        dict = {}
        while tmp:
            if tmp in dict:
                return tmp
            dict[tmp] = index
            index += 1
            tmp = tmp.next
        return None
        

File: LeetCode75_LevelOne/Linked_List_Cycle_II/Linked_List_Cycle_II.cc
// 142. Linked List Cycle II C++
#include <unordered_set>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        std::unordered_set<ListNode*> st;
        while (head) {
            if (st.find(head) != st.end()) return head;
            st.insert(head);
            head = head->next;
        }
        return nullptr;
    }
};
